File: tmGrammar/test/parser.cc
/**
 * @author      Takashi Matsushita
 * Created:     04 Jul 2014
 */

/** @todo */

/*====================================================================*
 * declarations
 *====================================================================*/
/*--------------------------------------------------------------------*
 * headers
 *--------------------------------------------------------------------*/
#include "tmUtil/tmUtil.hh"
#include "Algorithm.hh"
#include "Function.hh"
#include "Object.hh"


// ---------------------------------------------------------------------
// main
// ---------------------------------------------------------------------
int
main()
{
  const std::string indent = "  ";

  std::cout << "/////////////////////////////////////////////////////////\n\n";
  std::cout << "\t\tA logic parser ...\n\n";
  std::cout << "/////////////////////////////////////////////////////////\n\n";

  std::cout
    << "Give me a logic of the form :\n"
    << "  NOT (MU10 AND EG.gt.10p5+1) OR ((NOT TAU15[TAU-ISO_1] AND JET100[JET-ETA_1,JET-ETA_2]) AND (comb{MU10,MU20} AND NOT mass{MU10[MU-PHI_1,MU-ETA_1],EG10}[MASS_1,MASS_2]))\n";
  std::cout << "Type [q or Q] to quit\n\n";

  std::string str;
  while (getline(std::cin, str))
  {
    Algorithm::Logic::clear();

    if (str.empty() || str[0] == 'q' || str[0] == 'Q') break;

    bool rc = Algorithm::parser(str);
    if (not rc) continue;

    // analyse each token
    for (size_t ii = 0; ii < Algorithm::Logic::tokens.size(); ii++)
    {
      const std::string& token = Algorithm::Logic::tokens.at(ii);
      bool matched = false;

      // logical operator
      if (Algorithm::gateName.find(token) != Algorithm::gateName.end())
      {
        std::cout << token <<  " is a GATE" << std::endl;
        matched = true;
        continue;
      }

      // function
      for (Function::reserved::const_iterator cit = Function::functionName.begin();
           cit != Function::functionName.end(); cit++)
      {
        if (token.compare(0, cit->first.length(), cit->first) == 0)
        {
          std::cout << token <<  " is a function" << std::endl;

          Function::Item function;
          if (not Function::parser(token, function)) continue;
          std::cout << indent;
          function.print();
          matched = true;

          break;
        }
      }

      // object
      for (Object::reserved::const_iterator cit = Object::objectName.begin();
           cit != Object::objectName.end(); cit++)
      {
        if (token.compare(0, cit->first.length(), cit->first) == 0)
        {
          std::cout << token <<  " is an object" << std::endl;

          Object::Item object;
          if (not Object::parser(token, object)) continue;
          std::cout << indent;
          object.print();
          matched = true;

          break;
        }
      }

      if (matched) continue;
      TM_LOG_ERR("failed to parse '" << token << "'");
    }
  }

  std::cout << "Bye... :-) \n\n";
  return 0;
}
/* eof */

File: tmGrammar/test/parser.py
#!/usr/bin/env python

##
#  @author      Takashi Matsushita
#  Created:     07 Jul 2014
#

## @file parser.py demonstration of tmGrammar module functionality
## @todo


import tmGrammar

indent = "  "

print "/////////////////////////////////////////////////////////\n";
print "\t\tA logic parser ...\n";
print "/////////////////////////////////////////////////////////\n";

print "Give me a logic of the form :"
print "  NOT (MU10 AND EG.gt.10p5+1) OR ((NOT TAU15[TAU-ISO_1] AND JET100[JET-ETA_1,JET-ETA_2]) AND (comb{MU10,MU20} AND NOT mass{MU10[MU-PHI_1,MU-ETA_1],EG10}[MASS_1,MASS_2]))\n";

print "Type [q or Q] to quit\n";

while True:
  expression = raw_input("expression: ");
  if ((not expression) or (expression[0] in ['q', 'Q'])): break

  tmGrammar.Algorithm_Logic.clear()
  tmGrammar.Algorithm_parser(expression)

  for token in tmGrammar.Algorithm_Logic.getTokens():
    if token in tmGrammar.gateName:
      print token, " is a Gate"
      continue;

    isMatched = False

    for name in tmGrammar.functionName:
      if name == token[:len(name)]:
        print token, " is a function"
        isMatched = True
        o = tmGrammar.Function_Item()
        tmGrammar.Function_parser(token, o)
        o._print()
        break

    for name in tmGrammar.objectName:
      if name == token[:len(name)]:
        print token, " is an object"
        isMatched = True
        o = tmGrammar.Object_Item()
        tmGrammar.Object_parser(token, o)
        o._print()
        break

    if isMatched: continue
    print "failed to parse '", token, "'"

print "Bye... :-) \n"
#eof
